src/pm.c

```diff
@@ -0,0 +1,472 @@
+#include <stdarg.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+#define PM_COUNTER	"counter"
+#define PM_GAUGE	"gauge"
+#define PM_HISTOGRAM	"histogram"
+
+#define PM_LEN			128
+#define PM_LABEL_MAX		5
+#define PM_LABEL_VAL_MAX	5
+#define PM_METRIC_MAX		30
+#define PM_HIST_BIN_MAX		10
+
+typedef struct {
+	char		name[PM_LEN];
+	char		help[PM_LEN];
+	char		type[PM_LEN];
+	char		label[PM_LABEL_MAX][PM_LEN];
+	uint32_t	n_label;
+	double		bin[PM_HIST_BIN_MAX];
+	uint32_t	n_bin;
+	struct {
+		char	lblval[PM_LABEL_MAX][PM_LEN];
+		union {
+			double	 	counter;
+			double		gauge;
+			struct {
+				double		sum;
+				uint64_t	cnt[PM_HIST_BIN_MAX];
+				uint64_t	inf;
+			} histogram;
+		};
+	} s[PM_LABEL_VAL_MAX];
+	uint32_t	n_store;
```

Comment:
what exactly is n_store representing?

Reply:
the number of different permutation of value for the labels. For example

label[0]: "user", node"
---------------------
value[0]: "bob", "nyc1" <-- permutation 1
value[1]: "alice", "nyc2" <-- permutation 2
value[2]: "bog", "nyc2" <--- permutation 3

the key (label[N]) + (value[N]) point to a store[N]